### ISSUE TITLE
ipaddr_to_asc() added to 3.0.0 incorrectly formats IPv6 addresses

### ISSUE THREAD

**Issue by h-vn** [openssl/openssl#12695]
Function `ipaddr_to_asc` in `crypto/x509/v3_utl.c` converts a 4 or 16 octet value to an IPv4 or IPv6 address text represenation. What was noticed while testing Perl Net::SSLeay with OpenSSL 3.0.0 alpha5 and later is that `ipaddr_to_asc` text representation is incorrect with IPv6. The latest OpenSSL 1.1.1 and earlier work as expected, as shown below. The problem seems to have appeared during 3.0.0 development.

The example below is from an up-to-date macOS Catalina and Apple Xcode compilers.  I'd say this is not related to any particular platform.
```
% cc -v
Apple clang version 11.0.3 (clang-1103.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
% uname -a
Darwin imac-retina.local 19.6.0 Darwin Kernel Version 19.6.0: Sun Jul  5 00:43:10 PDT 2020; root:xnu-6153.141.1~9/RELEASE_X86_64 x86_64
```
This shows output with OpenSSL 1.1.1g that was compiled today:
```
% openssl version                                                                                            
OpenSSL 1.1.1g  21 Apr 2020
% openssl x509 -in t/data/testcert_extended.crt.pem -noout -text |grep -A1 'X509v3 Subject Alternative Name:'
            X509v3 Subject Alternative Name: 
                email:alt1@email.com, Registered ID:1.2.3.4.5, URI:http://name.dom.com/whatever, DNS:dns.name.com, IP Address:10.11.12.13, othername:<unsupported>, IP Address:2A00:1450:4001:C01:0:0:0:93
```
This shows output with OpenSSL master checked out and compiled today:
```
% openssl version                                                                                            
OpenSSL 3.0.0-alpha7-dev  (Library: OpenSSL 3.0.0-alpha7-dev )
% openssl x509 -in t/data/testcert_extended.crt.pem -noout -text |grep -A1 'X509v3 Subject Alternative Name:'
            X509v3 Subject Alternative Name: 
                email:alt1@email.com, Registered ID:1.2.3.4.5, URI:http://name.dom.com/whatever, DNS:dns.name.com, IP Address:10.11.12.13, othername: 1.2.3.4.5::othername, IP Address:2A0145400C010:
```
The [test certificate](https://github.com/radiator-software/p5-net-ssleay/blob/master/t/data/testcert_extended.crt.pem) from Perl Net::SSLeay is below:
```
-----BEGIN CERTIFICATE-----
MIIHGTCCBoKgAwIBAgIDEjRYMA0GCSqGSIb3DQEBBQUAMCsxCzAJBgNVBAYTAlVT
MQ4wDAYDVQQKEwVEZW1vMTEMMAoGA1UEAxMDQ0ExMB4XDTEyMDIwMTAwMDAwMFoX
DTM1MDMzMTIzNTk1OVowggHVMQswCQYDVQQGEwJVUzEOMAwGA1UECBMFU3RhdGUx
DTALBgNVBAcTBENpdHkxEDAOBgNVBAoTB0NvbXBhbnkxDTALBgNVBAsTBFVuaXQx
HTAbBgNVBAMTFGV4dGVuZGVkLnNlcnZlci5jZXJ0MSwwKgYJKoZIhvcNAQkBFh1h
bnllbWFpbEBleHRlbmRlZC5zZXJ2ZXIuY2VydDEVMBMGA1UECxMMQW5vdGhlciBV
bml0MRkwFwYDVQQLExBZZXQgQW5vdGhlciBVbml0MRUwEwYDVQQEEwxTdXJuYW1l
IHRleHQxEzARBgNVBAwTClRpdGxlIHRleHQxGTAXBgNVBA0TEERlc2NyaXB0aW9u
IHRleHQxEjAQBgNVBCkTCU5hbWUgdGV4dDEXMBUGA1UEKhMOR2l2ZW5OYW1lIHRl
eHQxFjAUBgNVBCsTDUluaXRpYWxzIHRleHQxFTATBgNVBC4TDGRuLnF1YWxpZmll
cjENMAsGA1UESBMEUm9sZTEdMBsGA1UELAwUZ2VuZXJhdGlvbl9xdWFsaWZpZXIx
FzAVBgNVBEETDlBzZXVkb255bSB0ZXh0MR0wGwYDVQQtDBR4NTAwX3VuaXFfaWRl
bnRpZmllcjCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEA2aWvhnencjZ7M/Ff
h2FyRejq29hF/XOSmJmmZgwirs9dnOq3kXrnx0y9XMBKPj1kdYW5omesX6pMIAyl
MNl5fADUtBQkxCneXnIcKAgvgxaqz1VqnTu3UBN4H/g82KOEkMw2mXEaY0TmvG3e
DgU9zvd/8T0ZUlTP6QmGA2OBkqcCAwEAAaOCA50wggOZMAwGA1UdEwEB/wQCMAAw
HQYDVR0OBBYEFKJ+nSndaOLx9HyMFO6u1yJFYi3/MFUGA1UdIwROMEyAFMgc2pIK
qUgIOnZ2FTgE8TTZFdAgoS+kLTArMQswCQYDVQQGEwJVUzEOMAwGA1UEChMFRGVt
bzExDDAKBgNVBAMTA0NBMYIDEjRWMAwGA1UdDwQFAwMH4oAwXgYDVR0lAQH/BFQw
UgYIKwYBBQUHAwEGCCsGAQUFBwMDBgorBgEEAYI3CgMBBggrBgEFBQcDBQYIKwYB
BQUHAwYGCCsGAQUFBwMHBggrBgEFBQcDDQYIKwYBBQUHAw4weAYDVR0RBHEwb4EO
YWx0MUBlbWFpbC5jb22IBCoDBAWGHGh0dHA6Ly9uYW1lLmRvbS5jb20vd2hhdGV2
ZXKCDGRucy5uYW1lLmNvbYcECgsMDaATBgQqAwQFoAsMCW90aGVybmFtZYcQKgAU
UEABDAEAAAAAAAAAkzBWBgNVHRIETzBNgRhpc3N1ZXIuYWx0bmFtZUBlbWFpbC5j
b22IAyoAAIYVaHR0cDovL2lzc3Vlci5jb20veHl6gg9pc3N1ZXIuYWx0Lm5hbWWH
BAECAwQwQwYDVR0fBDwwOjAboBmgF4YVaHR0cDovL3NlcnYxLmNvbS9jcmwxMBug
GaAXhhVodHRwOi8vc2VydjEuY29tL2NybDIwMAYIKwYBBQUHAQEEJDAiMCAGCCsG
AQUFBzABhhRodHRwOi8vb2NzcC93aGF0ZXZlcjARBglghkgBhvhCAQEEBAMCAFEw
KwYJYIZIAYb4QgECBB4WHGh0dHA6Ly9uZXRzY2FwZS5iYXNlLnVybC94enkwMQYJ
YIZIAYb4QgEDBCQWImh0dHA6Ly9uZXRzY2FwZS5yZXZvY2F0aW9uLnVybC94enkw
NAYJYIZIAYb4QgEEBCcWJWh0dHA6Ly9uZXRzY2FwZS5jYS5yZXZvY2F0aW9uLnVy
bC94enkwOgYJYIZIAYb4QgEHBC0WK2h0dHA6Ly9uZXRzY2FwZS5jZXJ0aWZpY2F0
ZS5yZW5ld2FsLnVybC94enkwMAYJYIZIAYb4QgEIBCMWIWh0dHA6Ly9uZXRzY2Fw
ZS5jYS5wb2xpY3kudXJsL3h6eTAeBglghkgBhvhCAQwEERYPc3NsLnNlcnZlci5u
YW1lMCUGCWCGSAGG+EIBDQQYFhZuZXRzY2FwZSBjb21tZW50IGZpZWxkMA0GCSqG
SIb3DQEBBQUAA4GBAKfdXUWwBvGjchb0VxVKspuvbpUnADOKcro1kdviRQRek46l
6Rjvq7vUZoUyeDOkFg08kxQQgNFOdnYLtuL6BkhGD7lVArM4s9anyp30umHgWEpR
0W6H0lfKI7nkc49sM7NJX0Srg4b6FVHhrNKhXC3ZPCOMIUqfW/Qi3ENJfVYM
-----END CERTIFICATE-----
```
Output with current master shows that `othername` value is `:othername`. It looks odd, but I think it's correct. The only problem I see is with IPv6 address text format.

I did not attempt a fix yet because I started thinking there could be something that, for example, converts a 16 octet value to something like ASN1_OCTET_STRING which could then be converted to IPv6 address text representation.

**Comment by mattcaswell** [comment]
`out += 3` just looks wrong here:

https://github.com/openssl/openssl/blob/2a33470b4f23bcf1cd66bbf645c855142efa0ed9/crypto/x509/v3_utl.c#L988-L994

We're writing potentially 5 characters of output, but incrementing 3 each time!? Also the output width is variable. Possible fixes are to use a fixed width format specifier, or to check the return value of BIO_snprintf to check what was actually written.

@DDvO

**Comment by levitte** [comment]
That loop was...  something.  I've a tentative fix in #12696, please try it out.
